src/PathOrderMonotonic.h

```diff
@@ -0,0 +1,449 @@
+//Copyright (c) 2021 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#ifndef PATHORDERMONOTONIC_H
+#define PATHORDERMONOTONIC_H
+
+#include <cmath> //For std::sin() and std::cos().
+#include <unordered_set> //To track starting points of monotonic sequences.
+#include <unordered_map> //To track monotonic sequences.
+
+#include "PathOrder.h"
+
+namespace cura
+{
+
+constexpr coord_t COINCIDENT_POINT_DISTANCE = 5; // In uM. Points closer than this may be considered overlapping / at the same place
+constexpr coord_t SQUARED_COINCIDENT_POINT_DISTANCE = COINCIDENT_POINT_DISTANCE * COINCIDENT_POINT_DISTANCE;
+
+/*!
+ * Class that orders paths monotonically.
+ *
+ * This is a utility class that changes the order in which things are printed,
+ * to ensure that they are printed in the same major direction. Printing
+ * adjacent lines in the same direction ensures that they layer on top of each
+ * other in the same way. That helps to make the entire surface look consistent.
+ *
+ * To use this class, first create an instance and provide some parameters as
+ * metadata. Then add polygons and polylines to the class. Then call the
+ * \ref optimize function to compute the order. Finally, print the polygons and
+ * polylines in the \ref paths field in the order in which they are given.
+ *
+ * In the output of this class, polylines and polygons are combined into a
+ * single list: \ref paths . Each path contains a pointer to the original
+ * polygon data, as well as whether that data represented a polygon or a
+ * polyline, which direction to print the path in, and where to start along the
+ * path.
+ *
+ * The monotonic order does not use the Z seam settings. It is meant to apply
+ * only to polylines. If given polygons, it will place the seam in the location
+ * closest to the source direction of the monotonicity vector.
+ */
+template<typename PathType>
+class PathOrderMonotonic : public PathOrder<PathType>
+{
+public:
+    using typename PathOrder<PathType>::Path;
+
+    PathOrderMonotonic(const AngleRadians monotonic_direction, const coord_t max_adjacent_distance, const Point start_point)
+    : monotonic_vector(std::cos(monotonic_direction) * 1000, std::sin(monotonic_direction) * 1000)
```

Comment:
Indeed, but the aim is not for the vector to be a certain physical distance long, so `MM2INT` is not used correctly here. The aim is just to get enough spatial resolution so that rounding errors from the `coord_t` type don't significantly alter the direction. If the vector had a length of 1, it could really only point in the 8 cardinal directions due to rounding to integers (and in fact it rounds down here, so probably you just get 0,0).

Ideally you'd make the other uses of the `1000` constant multiply by the length of the vector instead. But since that length is already constant and known it's really quite unnecessary.

I'll define a constant here to remove the duplication. Done here: a0656e670f3470197d3def1612186ec759478ebb

Reply:
I see. A constant is fine in that case, yes.